doctr/models/recognition/master/pytorch.py

```diff
@@ -204,10 +206,41 @@ def __init__(
     def make_mask(self, target: torch.Tensor) -> torch.Tensor:
         size = target.size(1)
         look_ahead_mask = ~ (torch.triu(torch.ones(size, size)) == 1).transpose(0, 1)[:, None]
-        target_padding_mask = ~ torch.eq(target, self.vocab_size + 2)  # Pad
+        target_padding_mask = ~ torch.eq(target, self.vocab_size + 2)  # Pad symbol
         combined_mask = target_padding_mask & look_ahead_mask
         return torch.tile(combined_mask.permute(1, 0, 2), (self.num_heads, 1, 1))
 
+    def compute_loss(
+        self,
+        model_output: torch.Tensor,
+        gt: torch.Tensor,
+        seq_len: torch.Tensor,
+    ) -> torch.Tensor:
+        """Compute categorical cross-entropy loss for the model.
+        Sequences are masked after the EOS character.
+
+        Args:
+            gt: the encoded tensor with gt labels
+            model_output: predicted logits of the model
+            seq_len: lengths of each gt word inside the batch
+
+        Returns:
+            The loss of the model on the batch
+        """
+        # Input length : number of timesteps
+        input_len = model_output.shape[1]
+        # Add one for additional <eos> token (sos disappear in shift!)
+        seq_len = seq_len + 1
+        # Compute loss: don't forget to shift gt! Otherwise the model learns to output the gt[t-1]!
+        # The "masked" first gt char is <sos>. Delete last logit of the model output.
+        cce = F.cross_entropy(model_output.permute(0, 2, 1)[:, :, :-1], gt[:, 1:], reduction='none')
```

Comment:
Two things:
- for the slicing, you can use ellipsis if this is a 3-D tensor: `input_t[:, :, :-1]` is the same as `input_t[..., :-1]` (basically the ellipsis will fill the remaining dimensions)
- better to slice before the permutation